### PR TITLE
Fix DOF in activeshader rendering, update camera parameters

### LINKED ISSUE #69: DOF not working in ActiveShade rendering

### PR DESCRIPTION
**Pull request by usakhelo**
Hi,
This should fix #69. 
Also fixes updating camera parameters during interactive rendering, such as FOV, aperture, blades and so on. Camera type cannot be changed so switching Enable Depth of Field checkbox won't change anything during the rendering.

Thanks,
Sergo.

### CODE REVIEW

**Comment by usakhelo** [comment]
Hi @dictoon,
Question on updating camera type - I've checked appleseed studio code and it seems that when I change camera type there, it removes old camera, creates new one and adds it to the `scene->cameras()` container. I've tried to do the same in max, but it always crashes. I see that studio schedules those changes to happen between `frame_end` and `render_begin`. Since there's no such things in max plugin yet, I've tried to pause the rendering, clear cameras(), add new camera, and then reinitialize and resume the rendering. It's still crashing. Do you think it is timing problem or some other things should also happen to update the camera? I was using `build_camera()` function to create new camera. And I think I can add simple task scheduling to render_controller and session_manager to make it work in the same manner as in studio. I was just wandering if I'm missing something more important.
Sorry for long question.
Thanks,
Sergo.

File: src/appleseed-max-impl/appleseedrenderer/projectbuilder.cpp [hunk 1]
```diff
@@ -1292,6 +1292,33 @@ namespace
...
+    const float film_width = camera_node->GetFilmWidth(time, FOREVER);
+    const float film_height = film_width * aspect;
+    params.insert("film_dimensions", asf::Vector2f(film_width, film_height));
+
+    // DOF settings.
+    params.insert("f_stop", camera_node->GetLensApertureFNumber(time, FOREVER));
+    params.insert("focal_distance", camera_node->GetFocusDistance(time, FOREVER));
+    switch (camera_node->GetBokehShape(time, FOREVER))
+    {
+    case MaxSDK::IPhysicalCamera::BokehShape::Circular:
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedrenderer/projectbuilder.cpp]: Half-indent cases.

File: src/appleseed-max-impl/appleseedrenderer/projectbuilder.h [hunk 2]
```diff
@@ -44,7 +44,9 @@
 
 // Forward declarations.
 namespace renderer { class Camera; }
+namespace renderer { class ParamArray; }
 namespace renderer { class Project; }
+
```
Comment (dictoon) [inline src/appleseed-max-impl/appleseedrenderer/projectbuilder.h]: Remove this blank line, these are all forward declarations.

**Comment by dictoon** [comment]
Regarding your question: What do you do exactly when you say "I've tried to pause the rendering"?

When the scene is edited in any way, there isn't much choice but _stopping_ rendering, reinitializing a bunch of things and restarting rendering, because entities may precompute and cache values based on their and other entities' parameters. So the general idea is to schedule a modification, return `ReinitializeRendering` from the renderer controller and apply modifications just before rendering starts again.

**Comment by usakhelo** [comment]
By pause I meant setting `RendererController->set_status(Status::PauseRendering)`. I didn't investigate on why exactly it crashed, but I'll check and try to implement scheduling and proper updating.

**Comment by dictoon** [comment]
Ah ok, `Status::PauseRendering` won't work, all it does is pause job execution, it doesn't do any reinitialization so entities may still hold a pointer to the old camera.

**Comment by dictoon** [comment]
Nice!

**Pull request merged**